Feat/ancestry sheets #12

zithith opened the pull request:
Basic Ancestry Sheet functionality added. Moved item tabs to be vertical on the right as per the actor sheets. Switched the editor to the built-in `<prose-mirror>` element.

Also fixed up the custom fonts.

TO DO:
- [x] Fix the editor value when populating for edit mode
- [x] Add advancement tab to ancestries
- [x] Update Culture sheet to account for general item changes (changes made but pending push)
- [x] Another style pass?
- [x] Remove custom editor component
- [x] check editing permissions (edit toggle for items/img and title always editable for right permissions?)
- [x] improve ancestry component within char sheet? (probably a separate issue)

stanavdb commented:
I'd intentionally left the tabs in the middle of the sheet instead of vertically on the right, as we don't have an actual design for the item sheets and I wanted to remain as generic as possible. Also, for the sake of commonality, this matches what 5e does.

stanavdb commented:
>     * [ ]  check editing permissions (edit toggle for items/img and title always editable for right permissions?)

I think here we follow what other systems do and have edit/view mode only for actor sheets (as they're interacted with a lot), and have item sheets always editable (if you have the right permissions).

zithith commented:
> I think here we follow what other systems do and have edit/view mode only for actor sheets (as they're interacted with a lot), and have item sheets always editable (if you have the right permissions).
Yeah, that's the route I would have wanted to go as well. Didn't realise it was a pattern, but recognised that it's just easier to work with sheets that are ready to edit when you open them.
I suppose there could be people who still want a "play" mode for items that editors can use to avoid making mistakes, but lets go with the easier option if that's not a standard

zithith commented:
>I'd intentionally left the tabs in the middle of the sheet instead of vertically on the right, as we don't have an actual design for the item sheets and I wanted to remain as generic as possible. Also, for the sake of commonality, this matches what 5e does.

OK, a bit of a google shows me that it's more prevalent than I thought (though I'm sure I've played at least one system that does the side tabs for non-actors). My own preference would be for the side tabs as they free up more space in the main window for content, plus it's internal consistency if all sheet tabs are the same, and it'd be more expandable. I'd honestly just thought that people left the horizontal tabs on things because they hadn't gotten round to changing it 🤣
But happy to revert the change, it's pretty simple.

stanavdb commented on src/system/applications/item/injury-sheet.ts:
```diff
@@ -50,11 +51,25 @@ export class InjuryItemSheet extends BaseItemSheet {
     public async _prepareContext(
         options: DeepPartial<foundry.applications.api.ApplicationV2.RenderOptions>,
     ) {
+        if (
+            this.item.system.description!.value ===
+            CONFIG.COSMERE.items.types.injury.desc_placeholder
+        ) {
+            this.item.system.description!.value = game.i18n!.localize(
+                this.item.system.description!.value!,
+            );
+        }
+
+        const enrichedDescValue = await TextEditor.enrichHTML(
+            this.item.system.description!.value!,
+        );
+
         return {
             ...(await super._prepareContext(options)),
             isPermanent:
                 this.item.system.type === InjuryType.PermanentInjury ||
                 this.item.system.type === InjuryType.Death,
+            descHtml: enrichedDescValue,
```
Can't we move the description enrichment into the base item?

zithith replied:
Done the refactor to make the description code common. I don't think we can move the localisation of the values into the data models due to the way the mixin process works, the i18n objects isn't populated on the game object when the mixins are initialised (if I'm following your thought correctly).
I did try and provide the ancestry DescriptionItemMixin call the localised string when I was first implementing the initial values bit as it seemed more sense to me to just do it cleanly there, but alas it didn't work.

zithith commented:
I was thinking there was a lot of repeated code and trying to come up with
a better approach after I'd turned the PC off.
That definitely sounds like a good plan to try!

On Sun, 29 Sept 2024, 08:28 stanavdb, ***@***.***> wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/system/applications/item/injury-sheet.ts
> <https://github.com/stanavdb/cosmere-rpg/pull/56#discussion_r1779954355>:
>
> > +        if (
> +            this.item.system.description!.value ===
> +            CONFIG.COSMERE.items.types.injury.desc_placeholder
> +        ) {
> +            this.item.system.description!.value = game.i18n!.localize(
> +                this.item.system.description!.value!,
> +            );
> +        }
> +
> +        const enrichedDescValue = await TextEditor.enrichHTML(
> +            this.item.system.description!.value!,
> +        );
> +
>          return {
>              ...(await super._prepareContext(options)),
>              isPermanent:
>                  this.item.system.type === InjuryType.PermanentInjury ||
>                  this.item.system.type === InjuryType.Death,
> +            descHtml: enrichedDescValue,
>
> Can't we move the description enrichment into the base item?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/stanavdb/cosmere-rpg/pull/56#pullrequestreview-2335753506>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ANGKWFPC3QRQR5OGB5TBZJTZY6TZDAVCNFSM6AAAAABPABUXUCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDGMZVG42TGNJQGY>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

zithith commented:
created issue #69 to cover the character sheet component aspect (and widened scope to be al item types)

zithith commented:
Final style touches done, will open another issue for theming.

stanavdb reviewed (commented):
While reviewing I ran into some small usability issues with the advancements. I figured I'd just do some suggested changes on a branch from this PR. If you like them you can merge them into the PR.

What I've changed:
1. **Free path is now a reference to an Item instead of an id.** We should use ids for things like prerequisites or other form of dependencies (so that GMs can modify/clone Items as they like but they'll still count as the "same"). However here we want to grant some path upon taking the ancestry, so we should refer to the Item directly.
2. **Free talents now uses references to Talent items instead of ids.** Same as point 1.
3. **Refactored talent picks to bonus talents and changed the way they're edited.** I've moved the restrictions to be per level, and changed up the way they're edited to be consistent with other Item sheets.

![image](https://github.com/user-attachments/assets/a47f82f6-1409-48df-8179-0d178a8e5f80)

zithith commented:
> While reviewing I ran into some small usability issues with the advancements. I figured I'd just do some suggested changes on a branch from this PR. If you like them you can merge them into the PR. https://github.com/stanavdb/cosmere-rpg/tree/ancestry-sheet-pr
> 
> What I've changed:
> 
> 1. **Free path is now a reference to an Item instead of an id.** We should use ids for things like prerequisites or other form of dependencies (so that GMs can modify/clone Items as they like but they'll still count as the "same"). However here we want to grant some path upon taking the ancestry, so we should refer to the Item directly.
> 2. **Free talents now uses references to Talent items instead of ids.** Same as point 1.
> 3. **Refactored talent picks to bonus talents and changed the way they're edited.** I've moved the restrictions to be per level, and changed up the way they're edited to be consistent with other Item sheets.
> 

OK, taken a look at that now.
1 & 2: That's great, I was thinking I would raise an issue to come back and update them to reference real objects at some point, but wasn't aware Items could hold collection within them (silly me, pretty sure I set something similar up using custom system builder a while back!) but whatever the approach it was going to take me another week or so to figure out how to plug it all together, so thought it better to do something quick for a 0.1 release (esp. as the advancement section was effectively just for reference, seemed like giving GMs a way to make notes quickly would have been better than my holding up the milestone). So you turning that round quickly is amazing! - side note, I'd love to pick your brains about the why/how you figured out what you needed to get embedded collections in. I can see *what* you added, but I don't 100% understand where it comes from 🤣

3: I like that it's refactored to support the embedded collections. That makes a lot of sense, and mechanically it's a lot simpler in places. And the styling and consistency is great. Display wise... I'm not a huge fan of lists-as-tables. I can see how logically it can make dealing with items in the TS files a bit cleaner, but I think it'd be better from an accessibility/screen-reader front to use table elements if you aren't a fan of the grid approach.

stanavdb commented:
> > While reviewing I ran into some small usability issues with the advancements. I figured I'd just do some suggested changes on a branch from this PR. If you like them you can merge them into the PR. https://github.com/stanavdb/cosmere-rpg/tree/ancestry-sheet-pr
> > What I've changed:
> > 
> > 1. **Free path is now a reference to an Item instead of an id.** We should use ids for things like prerequisites or other form of dependencies (so that GMs can modify/clone Items as they like but they'll still count as the "same"). However here we want to grant some path upon taking the ancestry, so we should refer to the Item directly.
> > 2. **Free talents now uses references to Talent items instead of ids.** Same as point 1.
> > 3. **Refactored talent picks to bonus talents and changed the way they're edited.** I've moved the restrictions to be per level, and changed up the way they're edited to be consistent with other Item sheets.
> 
> OK, taken a look at that now. 1 & 2: That's great, I was thinking I would raise an issue to come back and update them to reference real objects at some point, but wasn't aware Items could hold collection within them (silly me, pretty sure I set something similar up using custom system builder a while back!) but whatever the approach it was going to take me another week or so to figure out how to plug it all together, so thought it better to do something quick for a 0.1 release (esp. as the advancement section was effectively just for reference, seemed like giving GMs a way to make notes quickly would have been better than my holding up the milestone). So you turning that round quickly is amazing! - side note, I'd love to pick your brains about the why/how you figured out what you needed to get embedded collections in. I can see _what_ you added, but I don't 100% understand where it comes from 🤣
> 
> 3: I like that it's refactored to support the embedded collections. That makes a lot of sense, and mechanically it's a lot simpler in places. And the styling and consistency is great. I did have the restrictions per-level originally and then moved them up top because I wasn't sure how much busy work it would be to have to repeat the same or very nearly the same on each one... 🤷🏻‍♂️ Display wise... I'm not a huge fan of lists-as-tables. I can see how logically it can make dealing with items in the TS files a bit cleaner, but I think it'd be better from an accessibility/screen-reader front to use table elements if you aren't a fan of the grid approach.

The only real reason I haven't been using just regular tables is that they get a lot of default styling out of the box that we do not want.

zithith commented:
> The only real reason I haven't been using just regular tables is that they get a lot of default styling out of the box that we do not want.

Cool. I'll have a deep dive on this idea at some point this week. We can override a lot of the styling or potentially look at a sub-grid option to get the same row-container type approach. I might just merge the changes in tomorrow and then open a new issue/PR for the tables refactor, it'd likely be a new component or an overall approach I'd want to apply across any existing examples. But I'll also do more digging around what exactly is best practice these days (assuming we even can follow it within a foundry form!)

zithith commented:
* Merged in and addressed the conflicts with release branch.
* Issue #72 raised for the table debate
* [ ] Need to check items module.scss as I believe there are orphaned declarations for the old lists

stanavdb commented on src/style.scss:
```diff
@@ -67,36 +65,46 @@
 
+.application h1,
+span.document-name {
+    font-family: 'Penumbra Web Pro';
```
Small issue with this; It's hard to tell what letters are capitalised in document names. I think we can just add: `font-variant: small-caps;` as a fix.

zithith replied:
I did try and play around with this myself (bit of a sucker for small caps!). It looks good, but then I realised that the books just use the basic all caps approach so was keeping in line with that?

stanavdb replied:
I think we should use the small caps option here. As when editing the name it's impossible to see what letters you've got capitalised, but it does show in the items list on the side panel. It just kind of feels bad from a UX perspective.

zithith replied:
#74 raised for further discussion, have set the variant on the branch for now

stanavdb reviewed (changes requested):
We're almost there. I've highlighted a couple of minor issues. 
One more thing to think about: ATM you're applying the font globally, is that something we want? We'll want the font for our sheets but all the other Foundry element obviously weren't designed with it in mind. Might be better to keep the Foundry defaults for Foundry stuff and only use our fonts on our sheets.

zithith commented:
> ...ATM you're applying the font globally, is that something we want? We'll want the font for our sheets but all the other Foundry element obviously weren't designed with it in mind. Might be better to keep the Foundry defaults for Foundry stuff and only use our fonts on our sheets.

That's a fair point. Without rambling too much (and perhaps there lies the evidence that this should be a separate Discussion?), I have a personal preference when I see the foundry interface themed up. I wish there were more style themes around in the modules list (or even a separate section jus for themes, but I digress), and like it when a system takes the time to use the fact that Foundry was intentionally built on a basis that it can be easily themed from top to bottom. It improves immersion (or at least product familiarity) to me. And honestly I think the Laski sans font works quite well for the job, I'd not use Penumbra for UI elements and even hesitate to use the Palatin option (if we can find some way to provide that for the subheadings) for much beyond a few labels.
But you raise a good point about them not being designed for these fonts and we'd need to do thorough checking I guess. Perhaps we get a poll going about whether the themes should apply across the foundry base UI too? I say themes because as I write this I realise some/all of these font choices could well change between books...
So, OK, for now I'll just limit the fonts to anything `.application`/`.sheet`

zithith commented on src/system/applications/item/ancestry-sheet.ts:
```diff
@@ -0,0 +1,58 @@
+            window: {
+                resizable: true,
+                positioned: true,
+                title: '<i class="fas fa-timeline"></i> ' + this.name,
```
Good shout. I saw it in the API options and thought I'd try out some changes then forgot to remove when it didn't work.
I still want to try and address this as when minimised, each of the title bars for the item sheets tends to only be able to fit in the type name rather than the actual object name, which isn't helpful.
But someone will need to dig  bit deeper to see how we can override that base DocumentSheet behaviour then...

zithith replied:
#73 created for investigation

zithith replied:
removed current experiment

zithith commented on src/templates/item/components/talent-picks-list.hbs:
yeah, I picked up on the orphaned styles as mentioned above. I'll get on this too.

zithith replied:
removed

zithith commented on src/templates/item/partials/item-description-tab.hbs:
I *think* that's because I'm not passing in the editable parameter in the template. Probably just needs to be passed the root editable flag. Noted. And I'll do more thorough checking in future.

zithith replied:
This is apparently more complicated than I thought. Will investigate tomorrow

stanavdb commented:
> 
> That's a fair point. Without rambling too much (and perhaps there lies the evidence that this should be a separate Discussion?), I have a personal preference towards seeing the foundry interface themed up. I wish there were more style themes around in the modules list (or even a separate section jus for themes, but I digress), and like it when a system takes the time to use the fact that Foundry was intentionally built on a basis that it can be easily themed from top to bottom. It improves immersion (or at least product familiarity) to me. And honestly I think the Laski sans font works quite well for the job, I'd not use Penumbra for UI elements and even hesitate to use the Palatin option (if we can find some way to provide that for the subheadings) for much beyond a few labels. But you raise a good point about them not being designed for these fonts and we'd need to do thorough checking I guess. Perhaps we get a poll going about whether the themes should apply across the foundry base UI too? I say themes because as I write this I realise some/all of these font choices could well change between books... So, OK, for now I'll just limit the fonts to anything `.application`/`.sheet`

Yeah I'd like to take a look at complete foundry theming at some point, but that should be its own separate issue

zithith commented on src/templates/item/components/talents-list.hbs:
removed

zithith commented:
> > ... So, OK, for now I'll just limit the fonts to anything `.application`/`.sheet`
> 
> Yeah I'd like to take a look at complete foundry theming at some point, but that should be its own separate issue
Moved from global vars to just sheet styles (as per the h1s)

stanavdb commented on src/templates/item/partials/item-description-tab.hbs:
The description in a non-editable items (locked compendiums) is not getting displayed *entirely* correctly.
![image](https://github.com/user-attachments/assets/2894f410-8c19-49e0-8dd2-68082902c403)

I think we just need to add `flex-direction: column` to fix it.

zithith replied:
seems to have done the trick:
![image](https://github.com/user-attachments/assets/ace648af-8870-455f-82fb-bd2e7a753061)
